library/coretests/tests/mem.rs

```diff
@@ -795,3 +796,47 @@ fn const_maybe_uninit_zeroed() {
 
     assert_eq!(unsafe { (*UNINIT.0.cast::<[[u8; SIZE]; 1]>())[0] }, [0u8; SIZE]);
 }
+
+#[test]
+fn drop_guards_only_dropped_by_closure_when_run() {
+    let value_drops = Cell::new(0);
+    let value = DropGuard::new((), |()| value_drops.set(1 + value_drops.get()));
+    let closure_drops = Cell::new(0);
+    let guard = DropGuard::new(value, |_| closure_drops.set(1 + closure_drops.get()));
+    assert_eq!(value_drops.get(), 0);
+    assert_eq!(closure_drops.get(), 0);
+    drop(guard);
+    assert_eq!(value_drops.get(), 1);
+    assert_eq!(closure_drops.get(), 1);
+}
+
+#[test]
+fn drop_guard_into_inner() {
+    let dropped = Cell::new(false);
+    let value = DropGuard::new(42, |_| dropped.set(true));
+    let guard = DropGuard::new(value, |_| dropped.set(true));
+    let inner = DropGuard::into_inner(guard);
+    assert_eq!(dropped.get(), false);
+    assert_eq!(*inner, 42);
+}
+
+#[test]
+fn drop_guard_always_drops_value_if_closure_drop_unwinds() {
+    // Create a value with a destructor, which we will validate ran successfully.
+    let mut value_was_dropped = false;
+    let value_with_tracked_destruction = DropGuard::new((), |_| value_was_dropped = true);
+
+    // Create a closure that will begin unwinding when dropped.
+    let drop_bomb = DropGuard::new((), |_| panic!());
+    let closure_that_panics_on_drop = move |_| {
+        let _drop_bomb = drop_bomb;
+    };
+
+    // This will run the closure, which will panic when dropped. This should
+    // run the destructor of the value we passed, which we validate.
+    let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```

Comment:
If I'm understanding [this failure](https://github.com/rust-lang/rust/actions/runs/16423258546/job/46407148715?pr=144236) correctly, this line isn't actually catching the unwind the way it should.

Or maybe the test runner in CI mode is getting confused that we're unwinding and catching it. I'm not actually sure what is happening?

Reply:
I tried recreating this locally using `./x.py test library/coretests`, and I can't seem to. All tests are passing, and I'm not quite sure what's different.

Re-running all tests to see if that makes a difference.

Reply:
I can't reproduce locally (not as part of coretest, standalone). The panic still gets printed as we haven't installed a `panic_handler` but it should get caught.

Reply:
Nope, you're just testing on cranelift with `panic = "abort"`. Add `#[cfg(panic = "unwind")]` to the test.

Reply:
I don't often contribute directly to `rust-lang/rust`, so I appreciate the help!